internal/querynode/scheduler_policy.go

```diff
@@ -2,23 +2,305 @@ package querynode
+	return &userBasedTaskQueue{
+		cleanupTimestamp: time.Now(),
+		queue:            list.New(),
+		username:         username,
+	}
+}
+
+// User based task queue.
+type userBasedTaskQueue struct {
+	cleanupTimestamp time.Time
+	queue            *list.List
+	username         string
+}
+
+// Get length of user task
+func (q *userBasedTaskQueue) len() int {
+	return q.queue.Len()
+}
+
+// Add a new task to end of queue.
+func (q *userBasedTaskQueue) push(t readTask) {
+	q.queue.PushBack(t)
+}
+
+// Get the first task from queue.
+func (q *userBasedTaskQueue) front() readTask {
+	element := q.queue.Front()
+	if element == nil {
+		return nil
+	}
+	return element.Value.(readTask)
+}
+
+// Remove the first task from queue.
+func (q *userBasedTaskQueue) pop() {
+	front := q.queue.Front()
+	if front != nil {
+		q.queue.Remove(front)
+	}
+	if q.queue.Len() == 0 {
+		q.cleanupTimestamp = time.Now()
+	}
+}
+
+// Return true if user based task is empty and empty for d time.
+func (q *userBasedTaskQueue) expire(d time.Duration) bool {
+	if q.queue.Len() != 0 {
+		return false
+	}
+	if time.Since(q.cleanupTimestamp) > d {
+		return true
+	}
+	return false
+}
+
+// Merge a new task to task in queue.
+func (q *userBasedTaskQueue) mergeTask(rt readTask) bool {
+	for element := q.queue.Back(); element != nil; element = element.Prev() {
+		task := element.Value.(readTask)
+		if task.CanMergeWith(rt) {
+			task.Merge(rt)
+			return true
+		}
+	}
+	return false
+}
+
+// Implement user based schedule read policy.
+type userTaskPollingScheduleReadPolicy struct {
+	taskCount  int                   // task count in the policy.
+	route      map[string]*ring.Ring // map username to node of task ring.
+	checkpoint *ring.Ring            // last not schedule ring node, ring.Ring[list.List]
+}
+
+// Create a new user-based schedule read policy.
+func newUserTaskPollingScheduleReadPolicy() scheduleReadPolicy {
+	return &userTaskPollingScheduleReadPolicy{
+		taskCount:  0,
+		route:      make(map[string]*ring.Ring),
+		checkpoint: nil,
+	}
+}
+
+// Get length of task queue.
+func (p *userTaskPollingScheduleReadPolicy) len() int {
+	return p.taskCount
+}
+
+// Add a new task into ready task queue.
+func (p *userTaskPollingScheduleReadPolicy) addTask(rt readTask) {
+	username := contextutil.GetUserFromGrpcMetadata(rt.Ctx()) // empty user will compete on single list.
+	if r, ok := p.route[username]; ok {
+		// Add new task to the back of queue if queue exist.
+		r.Value.(*userBasedTaskQueue).push(rt)
+	} else {
+		// Create a new list, and add it to the route and queues.
+		newQueue := newUserBasedTaskQueue(username)
+		newQueue.push(rt)
+		newRing := ring.New(1)
+		newRing.Value = newQueue
+		p.route[username] = newRing
+		if p.checkpoint == nil {
+			// Create new ring if not exist.
+			p.checkpoint = newRing
+		} else {
+			// Add the new ring before the checkpoint.
+			p.checkpoint.Prev().Link(newRing)
+		}
+	}
+	p.taskCount++
```

Comment:
Is this taskCount concurrency safe?

Reply:
Schedule policy is always run in a single goroutine, no race condition here.